src/utils/FixedPointMathLib.sol

```diff
@@ -165,43 +165,46 @@ library FixedPointMathLib {
+            // each iteration instead of ~quadratically.
+            // The idea is to get z*z*y within a small factor of x.
+            // More iterations here gets y in a tighter range. Currently, we will have
+            // y in [256, 256*2^16). We ensure y>= 256 so that the relative difference
+            // between y and y+1 is small. If x < 256 this is not possible, but those cases
+            // are easy enough to verify exhaustively.
+            z := 181 // The 'correct' value is 1, but this saves a multiply later
             let y := x
-
-            // Find the lowest power of 2 that is at least sqrt(x).
-            if iszero(lt(y, 0x100000000000000000000000000000000)) {
-                y := shr(128, y) // Like dividing by 2 ** 128.
-                z := shl(64, z) // Like multiplying by 2 ** 64.
-            }
-            if iszero(lt(y, 0x10000000000000000)) {
-                y := shr(64, y) // Like dividing by 2 ** 64.
-                z := shl(32, z) // Like multiplying by 2 ** 32.
-            }
-            if iszero(lt(y, 0x100000000)) {
-                y := shr(32, y) // Like dividing by 2 ** 32.
-                z := shl(16, z) // Like multiplying by 2 ** 16.
+            // Note that we check y>= 2^(k + 8) but shift right by k bits each branch,
+            // this is to ensure that if x >= 256, then y >= 256.
+            if iszero(lt(y, 0x10000000000000000000000000000000000)) {
+                y := shr(128, y)
+                z := shl(64, z)
             }
-            if iszero(lt(y, 0x10000)) {
-                y := shr(16, y) // Like dividing by 2 ** 16.
-                z := shl(8, z) // Like multiplying by 2 ** 8.
+            if iszero(lt(y, 0x1000000000000000000)) {
+                y := shr(64, y)
+                z := shl(32, z)
             }
-            if iszero(lt(y, 0x100)) {
-                y := shr(8, y) // Like dividing by 2 ** 8.
-                z := shl(4, z) // Like multiplying by 2 ** 4.
+            if iszero(lt(y, 0x10000000000)) {
+                y := shr(32, y)
+                z := shl(16, z)
             }
-            if iszero(lt(y, 0x10)) {
-                y := shr(4, y) // Like dividing by 2 ** 4.
-                z := shl(2, z) // Like multiplying by 2 ** 2.
+            if iszero(lt(y, 0x1000000)) {
+                y := shr(16, y)
+                z := shl(8, z)
             }
-            if iszero(lt(y, 0x8)) {
-                // Equivalent to 2 ** z.
-                z := shl(1, z)
-            }
-
-            // Shifting right by 1 is like dividing by 2.
+            // Now, z*z*y <= x < z*z*(y+1), and y <= 2^(16+8),
+            // and either y >= 256, or x < 256.
+            // Correctness can be checked exhaustively for x < 256, so we assume y >= 256.
+            // Then z*sqrt(y) is within sqrt(257)/sqrt(256) of sqrt(x), or about 20bps.
+
+            // The estimate sqrt(x) = (181/1024) * (x+1) is off by a factor of ~2.83 both when x=1
```

Comment:
am i missing something or is it only off by ~0.17 at 1?

https://www.desmos.com/calculator/qxczllnaqm

Reply:
and a lot more at 256

Reply:
so e.g. at x=1, the estimate is 0.3535, which is ~1/2.83 of the correct value, i.e. off by a multiplicative factor of 2.83
I'll make this a bit clearer in a comment

Reply:
ahhhh gotcha my bad not familiar with that terminology